apps/docs/refactor.md

```diff
@@ -0,0 +1,565 @@
+```
+
+and the `index.js` file would be updated to explicitly export this component:
+
+**`src/stepper/index.js`**
+
+```js
+import Incrementor from './components/Incrementor';
+export Incrementor;
+```
+
+#### Using Shared State in a Feature Library ####
+
+In some circumstances, feature libraries may need to access shared state that is
+provided by another feature library. To make our discussion more concrete, let's
+assume that our stepper feature library needs to check if a user is logged in
+before allowing the step to be incremented. The login state of the application
+is provided by an authentication feature library.
+
+If the stepper library imports selectors directly from the authentication
+library to access the logged in state, then there is an implicit dependency
+created between the stepper library and the entry point that uses it because it
+assumes that at some point the authentication library has been configured
+properly by the entry point. This would be unintentional coupling.
+
+Instead, the entry point should be responsible for passing the configured
+authentication selectors directly to the stepper library via it's `configure`
+function. Here is how that might look:
+
+**`src/entry-points/some-page.js`**
+Now our entry point has to configure both the `stepper` and the `auth`
+libraries and manage how they access each other's shared state.
+
+```js
+import {createStore, combineReducers} from 'redux';
+import {* as auth} from '../auth';
+import {* as stepper} from '../stepper';
+
+const store = createStore(
+  combineReducers({
+    authState: auth.reducer,
+    stepperState: stepper.reducer
+  })
+);
+
+auth.configure({
+  getState: state => state.authState,
+});
+stepper.configure({
+  getState: state => state.stepperState,
+  authSelectors: auth.selectors,
+});
+```
+
+**`src/stepper/index.js`**
+
+Meanwhile, the configure function of our `stepper` library would now accept a
+set of auth selectors. It doesn't need to know where these selectors came from,
+which has a nice side effect of making the `stepper` library easier to test with
+a "fake" set of auth selectors.
+
+```js
+import reducer from './redux/reducer';
+import {INCREMENT_STEP} from './redux/actionTypes';
+import {incrementStep, setAuthSelector} from './redux/actionCreators';
+import {getStep, setRootSelector} from './redux/selectors';
+
+export reducer;
+export const actionTypes = {INCREMENT_STEP}
+export const actions = {incrementStep};
+export const selectors = {getStep};
+
+export function configure(config) {
+  setRootSelector(config.getState);
+  setAuthSelector(config.authSelectors);
+}
+```
+
+**`src/stepper/redux/actionCreators.js`**
+
+Now we can update our action creators to check the login state before actually
+dispatching the action.
+
+```js
+import * as actionTypes from './actionTypes';
+
+let authSelectors = {
+  isLoggedIn: () => throw new Error("Auth selectors not configured.");
```

Comment:
If we were being more formally OO I'd say the constructor is the right place to do that validation, since it _must_ happen when you're wiring everything together and it knows what can/should be injected.

Under the current design where our feature (stateful) libraries just have configure methods you could forget to call, I'm not sure.